Fixed exception during creation of Credit Memo when tax and discount applies to shipping price

vikalps4 opened the pull request:
Fixed issue  #19429
Fixed exception during creation of Credit Memo when tax and discount applies to shipping price

**Steps to reproduce**
1. Make an Order with shipping cost and VAT on the shipping.
2. Use a discount fixed on cart with an amount that can apply to shipping
3. Invoice and then create a credit memo from the invoice.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Updated Shipping.php

fixed issue  #19429
Credit Memo with discount get Exception on the Shipping

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=19570) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=19570) before we can accept your contribution.<br/>

magento-engcom-team commented:
Hi @vikalp-cedcoss. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

Commit:
Updated Shipping.php

sivaschenko commented on app/code/Magento/Sales/Model/Order/Creditmemo/Total/Shipping.php:
```diff
@@ -56,12 +56,12 @@ public function collect(\Magento\Sales\Model\Order\Creditmemo $creditmemo)
         $orderShippingInclTax = $order->getShippingInclTax();
         $orderBaseShippingInclTax = $order->getBaseShippingInclTax();
-        $allowedTaxAmount = $order->getShippingTaxAmount() - $order->getShippingTaxRefunded();
+        $allowedTaxAmount = $order->getShippingTaxAmount() + $order->getShippingDiscountTaxCompensationAmount() - $order->getShippingTaxRefunded();
```
Can you please break this line into several lines (as it is done on lines 61-64) to ensure it's not braking the 120 characters limit

vikalps4 replied:
Hi @sivaschenko 
I have updated the code as to ensure it's not braking the 120 characters limit.
Please review it now.

sivaschenko commented on app/code/Magento/Sales/Model/Order/Creditmemo/Total/Shipping.php:
```diff
@@ -56,12 +56,12 @@ public function collect(\Magento\Sales\Model\Order\Creditmemo $creditmemo)
         $orderShippingInclTax = $order->getShippingInclTax();
         $orderBaseShippingInclTax = $order->getBaseShippingInclTax();
-        $allowedTaxAmount = $order->getShippingTaxAmount() - $order->getShippingTaxRefunded();
+        $allowedTaxAmount = $order->getShippingTaxAmount() + $order->getShippingDiscountTaxCompensationAmount() - $order->getShippingTaxRefunded();
```
Can you please break this line into several lines (as it is done on lines 61-64) to ensure it's not braking the 120 characters limit

vikalps4 replied:
Hi @sivaschenko
I have updated it.

sivaschenko reviewed (changes requested):
@vikalp-cedcoss thanks for your contribution, can you please fix the code style issue described in the comment. It would be great also to cover the fixed functionality with an integration test.

Commit:
Updated Shipping.php

break the line into several lines to ensure it's not braking the 120 characters limit.

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-3623](https://jira.corp.magento.com/browse/ENGCOM-3623) has been created to process this Pull Request

magento-engcom-team commented:
@vikalp-cedcoss thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

sivaschenko commented:
Hi @vikalp-cedcoss we cannot reproduce the issue on 2.3-develop.

Manual testing scenario:

- Set Base Currency=Euro, Default Display Currency=Euro and Allowed Currencies=Euro
- Go to Stores->Configuration->Sales-Tax
- Set up taxes:
![image](https://user-images.githubusercontent.com/2028541/49578389-4c726500-f952-11e8-94e1-2a832427e313.png)
![image](https://user-images.githubusercontent.com/2028541/49578403-54320980-f952-11e8-9d1e-7e7e38dd68dd.png)

- Go to Stores->Configuration->Customers->Customer Configuration
- Set Create New Account Options->Show VAT Number on Storefront=Yes, Create New Account - Options->Show Tax/VAT Number=Required
- Go to Stores->Configuration->Sales->Shipping Settings->Origin, set Country=United Kingdom
- Go to Stores->Configuration->Sales->Shipping Methods->Flat Rate, set Type=Per Order, Price=7.9
- Create Customer with VAT Number=584451913, Country=United kingdom
- Create first product with price including tax is : 19.95 and second price including tax is: 26.90
- Create discount in Cart Price Rules:
Actions:
Apply=Fixed amount discount for whole cart
Discount Amount=5
Apply to Shipping Amount=Yes
- Log in as customer
- Create order with those products
- Create invoice and credit memo
![image](https://user-images.githubusercontent.com/2028541/49578424-63b15280-f952-11e8-8fad-d97a6aea157f.png)
 Credit memo is successfully created
![image](https://user-images.githubusercontent.com/2028541/49578447-7166d800-f952-11e8-9c68-41c27e3c9add.png)

Can you please provide more details for us to reporduce the original issue.

sidolov commented:
Hi @vikalp-cedcoss , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!
